api/app/signals/auth/tokens.py

```diff
@@ -34,19 +36,30 @@ def decode_claims(raw_claims):
         settings = get_settings()
         claims = loads(raw_claims)
 
-        user_id = None
-        try:
-            user_id = claims[settings['USER_ID_FIELD']]
-        except KeyError:
-            raise AuthenticationFailed("Field '{}' missing".format(settings['USER_ID_FIELD']))
+        # To temporarily support both authz tokens and KeyCloak tokens, we have
+        # to check the "sub" (authz) and "email" (KeyCloak and Dex) fields.
+        # Here we loop through the configured user id fields and check whether
+        # they are present and if so, if they contain an email address.
```

Comment:
The KeyCloak tokens contain both `sub` and `email`. The subject (`sub`) is a hash-like identifier, this is why I check that the field under consideration is a valid email address.

In case we have a `authz` token, there will not be a `email` claim. Our code, configured to look at both `sub` and `email` will only find `sub` and the is-this-a-valid-email-address test should succeed. In case we encounter a KeyCloak token (as currently configured) or likely (but we have to check) Dex token there will be both a `sub` claim and an `email` claim. Only the latter claim will be a valid email address.

I have looked at your proposed change and think we should discuss it.

Reply:
I did not know that KeyCloak had a `sub` and an `email` field. Let's talk a bit about my take on the function and see if we can combine the best of 2 worlds 😄